open-craft/opencraft
# Add features to support ecommerce and course discovery

In conjunction with open-craft/configuration#18, this pull request enables support for deploying an integrated ecommerce and course discovery environment.

Testing instructions:

Using this branch, create a new instance using the `opencraft-release/ficus.3` edx-platform branch and the configuration branch in open-craft/configuration#18. Include the following extra vars:

```yaml
DISCOVERY_ELASTICSEARCH_URL: "http://149.202.166.156:9200/"
SANDBOX_ENABLE_DISCOVERY: yes
SANDBOX_ENABLE_ECOMMERCE: yes
COMMON_HOSTNAME: "custom-local-hostname-here"
DISCOVERY_VERSION: "c96c77bbf80650b3d3cc238335cc5205e24b775f"
```

Currently, we don't support Elasticsearch as a database backend, so we have to pass that in manually. Additionally, discovery and ecommerce still aren't first-class citizens, so they must be enabled via extra var. We need to set the hostname, other than the FQDN, that the appserver refers to itself as so that we can properly go out to the load balancer-terminated SSL connection. And finally, newer versions of course discovery don't work properly with OIDC login.

Once the spawn is complete, you'll need to take the following steps to finish setup (these are one-time actions that are stored in the database):

1. Update the staff user so that a first name is associated with it (some routines look for a user profile).
2. Update the database-backed OAuth2 clients for course discovery and ecommerce so that they're associated with the staff user.
3. In the sandbox ecommerce env, configure the Enterprise Partner, Site, and SiteConfiguration [per the instructions on this page](http://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/ecommerce/install_ecommerce.html#configure-a-site-partner-and-site-configuration).
4. In the sandbox discovery env, configure a partner [as described here](https://open-edx-course-catalog.readthedocs.io/en/latest/getting_started.html#configure-partners).
5. If using a load balancer that does not fetch real certs by default, ssh into that load balancer and use a set of commands that looks something like this to get real certificates for your site:
    ```bash
    sudo letsencrypt certonly --email ops@opencraft.com --authenticator standalone --standalone-supported-challenges http-01 --http-01-port 8080 --force-renewal -d haikuginger-ecom2.plebia.net -d studio-haikuginger-ecom2.plebia.net -d ecommerce-haikuginger-ecom2.plebia.net -d discovery-haikuginger-ecom2.plebia.net
    sudo service haproxy restart
    ```
6. You should now be able to hit the `/login` endpoint on both discovery and ecommerce and have the OIDC flow log you in once you've authenticated with the LMS.
7. In the sandbox discovery env, run the following command and verify that it finishes without incident:
    ```bash
    ./manage.py refresh_course_metadata
    ```
